### PR TITLE
PR revisions

### PR DESCRIPTION
**Pull request by jbaublitz**
Resolves comments in #13

### CODE REVIEW

**Comment by jbaublitz** [comment]
@ojeda @wedsonaf Any last comments on this?

File: parrot.rs [hunk 1]
```diff
@@ -1,73 +1,62 @@
...
+struct ParrotSafe(Pin<Box<Registration<ParrotOps>>>);
 
 impl kernel::Module for ParrotSafe {
-    fn init(name: &'static CStr, module: &'static ThisModule) -> Result<Self> {
-        let mut registration = Registration::new_pinned(name, 0, module)?;
-        registration.as_mut().register::<ParrotOps>()?;
-        Ok(ParrotSafe(registration))
+    fn init(_: &'static CStr, _: &'static ThisModule) -> Result<Self> {
+        Ok(ParrotSafe(Registration::new_pinned(
+            format_args!("parrot"),
```
Comment (wedsonaf) [inline parrot.rs]: Not that it matters much, but we have a shorter macro for this declared in the prelude: `fmt`

File: parrot.rs [hunk 2]
```diff
@@ -1,73 +1,62 @@
...
-        ioctl: false,
-        mmap: false,
-        poll: false,
-        seek: false,
-    };
+struct ParrotOps;
 
-    type Data = Box<AtomicUsize>;
+impl Operations for ParrotOps {
+    declare_file_operations!(read);
```
Comment (wedsonaf) [inline parrot.rs]: We merged `#[vtable]` support (also included in the prelude), so you probably want to update the code to use it as `declare_file_operations!` has been removed.

File: parrot.rs [hunk 3]
```diff
@@ -1,73 +1,62 @@
...
+            return Err(EINVAL);
+        }
+        let (frame, frame_offset) = calc_frame_and_offset(offset);
+        let frame = FRAMES.get(frame).ok_or(EIO)?;
+        let offset_usize: usize = frame_offset.try_into()?;
+        let s = if buf.len() >= frame.len() - offset_usize {
+            &frame.as_bytes()[offset_usize..]
+        } else {
+            &frame.as_bytes()[offset_usize..offset_usize + buf.len()]
+        };
```
Comment (wedsonaf) [inline parrot.rs]: Untested, but perhaps you could write something like the expression below for this?

```rust
let s = &frame.as_bytes()[offset_usize..][..core::cmp::min(buf.len(), frame.len() - offset_usize)];
```

File: parrot.rs [hunk 4]
```diff
@@ -1,73 +1,62 @@
...
+        buf.write_slice(s)?;
+        if offset_usize + s.len() == frame.len() {
+            // # Safety: This invocation of `msleep()` takes a valid millisecond
```
Comment (wedsonaf) [inline parrot.rs]: Nit: the convention is to prefix the comment with `SAFETY:`; the `# Safety` format is used when documenting the safety requirements for `unsafe` functions and traits.

**Review by wedsonaf — COMMENTED** [review]
Thanks for making the changes! I really like the byte-based offset of the frames.

File: parrot.rs (hunk 3 above)
Comment (jbaublitz) [reply parrot.rs]: I had tried that earlier when I was testing code that was attempting to inject null bytes at the end. This turned out to be pretty unnecessary and really error prone so I removed the `min()` but I think it makes more sense here now.

File: parrot.rs (hunk 4 above)
Comment (jbaublitz) [reply parrot.rs]: Whoops! I read both the part that comments should be in markdown and the part about `SAFETY:` but only the markdown mention stuck. I'll fix this.

**Pull request merged**